shaping/spacing.go

```diff
@@ -0,0 +1,112 @@
+package shaping
+
+import (
+	"github.com/go-text/typesetting/di"
+	"golang.org/x/image/math/fixed"
+)
+
+// AddWordSpacing alters the run, adding [additionalSpacing] on each
+// word separator.
+// [text] is the input slice used to create the run.
+//
+// See also https://www.w3.org/TR/css-text-3/#word-separator
+func (run *Output) AddWordSpacing(text []rune, additionalSpacing fixed.Int26_6) {
+	isVertical := run.Direction.IsVertical()
+	for i, g := range run.Glyphs {
+		// find the corresponding runes :
+		// to simplify, we assume the words separators are not produced by ligatures
+		// so that the cluster "rune-length" is 1
+		if g.RuneCount != 1 {
+			continue
+		}
+		r := text[g.ClusterIndex]
+		switch r {
+		case '\u0020', // space
+			'\u00A0',                   // no-break space
+			'\u1361',                   // Ethiopic word space
+			'\U00010100', '\U00010101', // Aegean word separators
+			'\U0001039F', // Ugaritic word divider
+			'\U0001091F': // Phoenician word separator
+		default:
+			continue
+		}
+		// we have a word separator: add space
+		// we do it by enlarging the separator glyph advance
+		// and distributing space around the glyph content
+		if isVertical {
+			run.Glyphs[i].YAdvance += additionalSpacing
+			run.Glyphs[i].YOffset += additionalSpacing / 2
+		} else {
+			run.Glyphs[i].XAdvance += additionalSpacing
+			run.Glyphs[i].XOffset += additionalSpacing / 2 // distribute space around the char
+		}
```

Comment:
I think this code makes the assumption that the cluster contains only one glyph mapping to the space rune. While I hope this is true for all reasonable fonts, I suppose it isn't necessarily true. We should do one of:

- ignore multi-glyph clusters in this loop
- distribute the space we want to add across all glyphs in the cluster

The first option is much simpler. :D